src/libostree/ostree-fetcher-soup.c

```diff
@@ -1072,11 +1072,24 @@ on_request_sent (GObject        *object,
               GIOErrorEnum code;
```

Comment:
One thing that feels odd to me is that the way this is structured, we'll be retrying the whole set of mirrors if the *last one* on the list returns a transient error. So e.g. the first N-1 mirrors could return `NOT_FOUND` and if the Nth mirror times out, we'll be retrying all N mirrors, right? Whereas I think what we want is to retry only the Nth mirror. It seems like fixing this properly would probably require indeed teaching each backend-specific fetcher about retries.

Anyway, I think it's definitely worth getting the current approach in and iterate on that since AFAIK mirrorlists are not widely used yet.

Reply:
I’d definitely prefer to get the current approach in first. That said, I’m not sure whether I will have time afterwards to work on improvements to mirror handling (I’m working on this to fix a problem with our Endless downloads timing out, and mirrors don’t really impact on that).

However, I’m not entirely convinced about the situation you’re considering. Are we ever expecting the majority of mirrors to return `NOT_FOUND`? Sounds like an unlikely corner case to me. Even so, `OstreeFetcherSoup` maintains state about its mirrors, so I think the problem could be addressed by storing a bit more state there, rather than by rearchitecting things. Specifically, instead of storing a single `mirrorlist_idx` pointing to the best mirror to try, store some flags for each mirror, and track whether each mirror has returned `NOT_FOUND`, or returned a transient error, at all. Prioritise requesting from mirrors which haven’t. (This could also be implemented by storing a priority for each mirror, and demoting it on `NOT_FOUND` or transient errors.)

Reply:
To clarify, I wasn't suggesting you would fix those up. Just something that'd need to be done down the road. As I said, I'm not aware of any big users of mirrorlists yet, so it's not a pressing concern.

> Are we ever expecting the majority of mirrors to return `NOT_FOUND`? Sounds like an unlikely corner case to me.

Heh, sure. The `NOT_FOUND` was an example; anything "non-transient" would work. Anyway, I do agree with you that it could be done without lowering the retry logic down a level.